virtcontainers/sandbox.go

```diff
@@ -1334,11 +1349,43 @@ func togglePauseSandbox(sandboxID string, pause bool) (*Sandbox, error) {
 func (s *Sandbox) HotplugAddDevice(device api.Device, devType config.DeviceType) error {
 	switch devType {
 	case config.DeviceVFIO:
-		vfioDevice, ok := device.(*drivers.VFIODevice)
+		vfioDevices, ok := device.GetDeviceDrive().([]*config.VFIODrive)
 		if !ok {
 			return fmt.Errorf("device type mismatch, expect device type to be %s", devType)
 		}
-		return s.hypervisor.hotplugAddDevice(*vfioDevice, vfioDev)
+		addedDev := []*config.VFIODrive{}
+		var err error
+		defer func() {
+			// if err happens,roll back and remove added device!
+			if err != nil {
+				for _, dev := range addedDev {
+					if rollbackErr := s.hypervisor.hotplugRemoveDevice(dev, vfioDev); rollbackErr != nil {
+
+						s.Logger().WithError(rollbackErr).
+							WithFields(logrus.Fields{
+								"sandboxid":       s.id,
+								"vfio device ID":  dev.ID,
+								"vfio device BDF": dev.BDF,
+							}).Error("failed to remove vfio device for rolling back")
+					}
+				}
+			}
+		}()
+
+		// adding a group of VFIO devices
+		for _, dev := range vfioDevices {
+			if err = s.hypervisor.hotplugAddDevice(dev, vfioDev); err != nil {
+				s.Logger().WithError(err).
```

Comment:
You could define a new logger variable here to avoid the `WithFields()` duplication:

```
logger := s.Logger().WithFields(...)

defer func () {
    :
    logger.WithError(rollbackErr).Error("failed to remove vfio device for rolling back")
   :
}()

:

logger.WithError(err).Error("failed to hotplug VFIO device")
```

Same comment for code below.

Reply:
This is great, thank!

Reply:
Oops, this is in for loop, and I need to display the detailed information for each dev, so I'm afraid a logger template won't be good enough.